### PR TITLE
Add spec load pass to check for undefined build args

### LINKED ISSUE #274: [BUG] Argument substitution should handle undefined args used for substitution
### Expected Behavior

Given some substituion like `$FOO`, if `$FOO` is not defined dalec should error out similar to how bash with `set -u` does for unknown vars.
This can help prevent user errors (e.g. mispells) from going unnoticed.

### Actual Behavior

Currently this will give you an empty string when there is no matching substitution.

### Steps To Reproduce

_No response_

### Are you willing to submit PRs to contribute to this bug fix?

- [ ] Yes, I am willing to implement it.

### PR DESCRIPTION
**Pull request by adamperlin**
**What this PR does / why we need it**:
Currently, if a build arg is not specified in the `args` section of a Dalec spec file, its value is set to `""` when the spec is loaded. This is confusing behavior, as a user could easily forget to declare a build-arg or mistype a build-arg reference, leading to an empty string substitution which could be hard to track down.

This PR adds an additional pass upon loading the spec which tokenizes all the build args first, and checks to ensure that each referenced arg does appear in the `args` section of the spec, raising an error if not. 

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #274 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by cpuguy83** [comment]
Did you happen to look into using https://pkg.go.dev/github.com/moby/buildkit/frontend/dockerfile/shell?utm_source=godoc#Lex.ProcessWordWithMatches instead?
I believe that should give us the list of vars which weren't in our env list.

**Comment by adamperlin** [comment]
Ah no, I was not aware that function existed. I will see if I can reimplement using that instead

**Comment by adamperlin** [comment]
It looks like the version of this function you referenced exists in a newer version of buildkit (0.14.0) than the one we import in dalec (0.13.1). Are we ok to bump the buildkit dependency?

**Comment by cpuguy83** [comment]
Oh nice, yes we can/should bump it.

File: load.go [hunk 1]
```diff
@@ -38,67 +38,75 @@ func platformArg(key string) bool {
...
 const DefaultPatchStrip int = 1
 
-func (s *Source) substituteBuildArgs(args map[string]string) error {
+func expandArgs(lex *shell.Lex, s string, args map[string]string) (string, *ErrorList) {
+	result, err := lex.ProcessWordWithMatches(s, args)
+	if err != nil {
+		return "", AsList(err)
+	}
+
+	var errL ErrorList
```
Comment (cpuguy83) [inline load.go:47]: You can use `errors.Join` to do this without needing a separate type.

File: load.go [hunk 2]
```diff
@@ -471,22 +490,22 @@ func (c *TestSpec) processBuildArgs(lex *shell.Lex, args map[string]string, name
...
-		step.Stderr = stderr
 
+		step.Stderr = stderr
 		c.Steps[i] = step
 	}
 
 	for name, f := range c.Files {
 		if err := f.processBuildArgs(lex, args); err != nil {
-			return errors.Wrap(err, name)
+			errL.Append(errors.Wrap(err, name))
```
Comment (cpuguy83) [inline load.go:508]: The error is never returned in this case now.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: load.go (hunk 2 above)
Comment (adamperlin) [reply load.go:508]: Thank you for catching this!

File: load.go [hunk 3]
```diff
@@ -44,67 +44,96 @@ func platformArg(key string) bool {
...
+func expandArgs(lex *shell.Lex, s string, args map[string]string) (string, error) {
+	result, err := lex.ProcessWordWithMatches(s, args)
+	if err != nil {
+		return "", err
+	}
+
+	var errs []error
+	for m := range result.Unmatched {
+		if !knownArg(m) {
+			errs = append(errs, fmt.Errorf(`build arg "%s" not declared`, m))
```
Comment (cpuguy83) [inline load.go:56]: Nit, this can be a continue here to remove the `else if` to just an `if`

**Review by cpuguy83 — CHANGES_REQUESTED** [review]
MInor readability nit, but otherwise LGTM

**Review by cpuguy83 — APPROVED** [review]

**Pull request merged**